rfc/rfc-0005-customizable-consumption-offset.md

```diff
@@ -0,0 +1,68 @@
+# RFC-0005: Optional starting offset for processor inputs
+
+**Authors:** Eric Bottard
+
+**Status:**
+
+**Pull Request URL:** https://github.com/projectriff/riff/pull/1367
+
+**Superseded by:** N/A
+
+**Supersedes:** N/A
+
+**Related:** rfc-0003: developer utils, which also support viewing stream data from the earliest offset.
+
+
+## Problem
+Sometimes users want to replay data through functions, maybe because the function logic has changed, or maybe because downstream systems have lost state and that state needs to be re-created. The current Processor concept in riff doesn't support such a mechanism though, and always consumes data from the (current) end of streams.
+
+### Anti-Goals
+
+## Solution
+With a top to bottom view of the user experience, the proposed solution is to
+- introduce an offset option on each input binding when a processor is created
+- persist that option in the Processor CRD instance on the cluster
+- have the Processor reconciler forward that option to the streaming-processor, so that it changes its SubscriptionRequests accordingly
+
+### User Impact
+This RFC proposes to change the `Inputs` field of the `Processor` CRD from
+```go
+Inputs []StreamBinding `json:"inputs"`
+```
+to
+```go
+Inputs []InputStreamBinding `json:"inputs"`
+...
+type InputStreamBinding struct {
+    Stream string `json:"stream"`
+    Alias string `json:"alias,omitempty"`
+
+	// Where to start consuming this stream the first time a processor runs.
+    StartOffset string `json:"startOffset"`
+}
+```
+
+with `StartOffset` supporting only two special, textual values for now: `earliest` and `latest` (default). Future support for parsing to a non negative number after failing to recognize those two values is possible.
+
+This RFC proposes to change the `riff streaming processor create` command from
+```
+riff streaming processor create <name> ... --input [alias:]stream ...
+```
+to
+```
+riff streaming processor create <name> ... --input [alias:]stream[:offset] ...
```

Comment:
Can we reliably distinguish a value like `foo:latest`? Is `foo` the alias or stream, is `latest` the stream or offset?

While not perfect, making the textual values `Earliest` and  `Latest` would make distinction possible since the stream name can't contain uppercase characters.

Reply:
Very good point

Reply:
I don't think we should caps as a workaround for this. Maybe we're trying to include too much information in the same string?

Reply:
or maybe we use `[alias:]stream[@offset]` instead?

Reply:
`@` is very good too, but reads better with strings like start/end than with earliest/latest.

Reply:
thinking "start reading the stream AT the earliest offset" works for me :-)

Reply:
+1 for `@` as the delimiter rather than capitalizing `earliest` and `latest`.